File: database-driver-plugin/src/main/java/de/lino/database/database/nosql/rethinkdb/RethinkDBDatabaseProvider.java
package de.lino.database.database.nosql.rethinkdb;

import com.rethinkdb.RethinkDB;
import com.rethinkdb.gen.ast.Db;
import com.rethinkdb.net.Connection;
import com.rethinkdb.net.Result;
import de.lino.database.database.AbstractCachedDatabaseSection;
import de.lino.database.database.AbstractLazyDatabaseProvider;
import de.lino.database.database.SectionConfig;
import de.lino.database.database.auth.Credentials;
import de.lino.database.database.DatabaseProvider;
import de.lino.database.database.DatabaseSection;
import lombok.Getter;
import org.jetbrains.annotations.NotNull;

import java.util.function.Consumer;

/**
 * The {@link DatabaseProvider} backed by a RethinkDB database, each {@link DatabaseSection} a
 * table via {@link RethinkDBDatabaseSection}, all sharing this database's single
 * {@link Connection}. Section lifecycle and caching live in {@link AbstractLazyDatabaseProvider};
 * this class only supplies the table-level storage operations - listing tables, constructing a
 * {@link RethinkDBDatabaseSection}, dropping a table. RethinkDB's {@link Connection} multiplexes
 * concurrent queries over one underlying socket and is itself thread-safe, so every method here
 * is safe to call concurrently without additional locking.
 */
@Getter
public class RethinkDBDatabaseProvider extends AbstractLazyDatabaseProvider {

    /**
     * The connection shared by this database and every {@link RethinkDBDatabaseSection} it creates.
     */
    private final Connection connection;

    /**
     * The database this database is connected to.
     */
    private final Db db;

    /**
     * Connects to a RethinkDB database with {@code credentials} and discovers every existing
     * table's name. Only names - no section objects, no rows - so construction cost is one
     * {@code tableList} query, independent of how much the database holds.
     *
     * @param credentials the login credentials and connection details to connect with
     */
    public RethinkDBDatabaseProvider(@NotNull Credentials credentials) {

        this.connection = RethinkDB.r.connection()
                .hostname(credentials.getAddress())
                .port(credentials.getPort())
                .user(credentials.getUserName(), credentials.getPassword())
                .db(credentials.getDatabase())
                .connect();
        this.db = RethinkDB.r.db(credentials.getDatabase());

        this.reload();

    }

    @Override
    public void shutdown() {
        this.connection.close();
        this.forgetSections();
    }

    /**
     * {@inheritDoc}
     * <p>
     * Runs the database's {@code tableList} query, streaming each table name to
     * {@code consumer}.
     */
    @Override
    protected void discoverNames(@NotNull Consumer<String> consumer) {

        try (final Result<String> names = this.db.tableList().run(this.connection, String.class)) {
            names.forEach(consumer);
        }

    }

    @Override
    protected AbstractCachedDatabaseSection constructSection(@NotNull String name, @NotNull SectionConfig config) {
        return new RethinkDBDatabaseSection(name, this.connection, this.db, config);
    }

    @Override
    protected void dropSectionRemote(@NotNull String name) {
        this.db.tableDrop(name).run(this.connection);
    }

}

File: python/database-driver-plugin/src/database_driver/plugin/database/nosql/rethinkdb/rethinkdb_database_provider.py
"""Mirror of ``de.lino.database.database.nosql.rethinkdb.RethinkDBDatabaseProvider``."""

from __future__ import annotations

from collections.abc import Callable

from database_driver.api.database.auth.credentials import Credentials
from database_driver.api.database.section_config import SectionConfig
from database_driver.plugin.database.abstract_cached_database_section import AbstractCachedDatabaseSection
from database_driver.plugin.database.abstract_lazy_database_provider import AbstractLazyDatabaseProvider
from database_driver.plugin.database.nosql.rethinkdb.rethinkdb_database_section import RethinkDBDatabaseSection


class RethinkDBDatabaseProvider(AbstractLazyDatabaseProvider):
    """The ``DatabaseProvider`` backed by a RethinkDB database, each section a table via
    ``RethinkDBDatabaseSection``, all sharing this database's single connection. Section
    lifecycle and caching live in ``AbstractLazyDatabaseProvider``; this class only
    supplies the table-level storage operations - listing tables, constructing a
    section, dropping a table.

    Connects with ``credentials`` and discovers every existing table's name. Only names
    - no section objects, no rows - so construction cost is one ``table_list`` query,
    independent of how much the database holds.
    """

    def __init__(self, credentials: Credentials) -> None:
        super().__init__()

        from rethinkdb import RethinkDB

        self.r = RethinkDB()
        self.connection = self.r.connect(
            host=credentials.address,
            port=credentials.port,
            user=credentials.user_name,
            password=credentials.password,
            db=credentials.database,
        )
        self.db = self.r.db(credentials.database)

        self.reload()

    def shutdown(self) -> None:
        self.connection.close()
        self.forget_sections()

    def discover_names(self, consumer: Callable[[str], None]) -> None:
        """Runs the database's ``table_list`` query, streaming each table name to
        ``consumer``."""
        for name in self.db.table_list().run(self.connection):
            consumer(name)

    def construct_section(self, name: str, config: SectionConfig) -> AbstractCachedDatabaseSection:
        return RethinkDBDatabaseSection(name, self.r, self.connection, self.db, config)

    def drop_section_remote(self, name: str) -> None:
        self.db.table_drop(name).run(self.connection)
